forge/test/operators/pytorch/tm/test_reshape.py

```diff
@@ -0,0 +1,361 @@
+
+from forge.verify.config import VerifyConfig
+
+from forge.verify.value_checkers import AllCloseValueChecker
+from forge.verify.verify import verify as forge_verify
+
+from test.operators.utils import InputSourceFlags, VerifyUtils
+from test.operators.utils import InputSource
+from test.operators.utils import TestVector
+from test.operators.utils import TestPlan
+from test.operators.utils import TestPlanUtils
+from test.operators.utils import FailingReasons
+from test.operators.utils.compat import TestDevice
+from test.operators.utils import TestCollection
+from test.operators.utils import TestCollectionCommon
+from test.operators.utils import ValueRanges
+
+from test.operators.pytorch.eltwise_unary import ModelFromAnotherOp, ModelDirect, ModelConstEvalPass
+
+
+def prime_factors(n):
+    """Return the prime factors of a given number."""
+    factors = []
+    while n % 2 == 0:
+        factors.append(2)
+        n //= 2
+    
+    for i in range(3, int(math.sqrt(n)) + 1, 2):
+        while n % i == 0:
+            factors.append(i)
+            n //= i
+    
+    if n > 2:
+        factors.append(n)
+    
+    return factors
+
+
+def generate_random_shape(source_shape, max_dims=4):
+    """
+    Generates a random shape with the same volume as the source shape.
+    Args:
+        source_shape (tuple): The original shape of the tensor.
+        max_dims (int, optional): The maximum number of dimensions for the generated shape. Defaults to 4.
+    Returns:
+        tuple: A randomly generated shape with the same volume as the source shape.
+    Raises:
+        ValueError: If the volume or number of dimensions is not a positive integer.
+        ValueError: If the generated shape does not match the desired volume.
+    Notes:
+        - The function ensures that the product of the dimensions of the generated shape equals the product of the dimensions of the source shape.
+        - If the generated shape is the same as the source shape, an additional dimension with size 1 is appended to the shape.
+    """
+
+    rng = random.Random(math.prod(source_shape) + 1)
+    
+    vol = math.prod(source_shape)
+    num_dims = rng.randint(1, max_dims)
+    
+    if vol <= 0 or num_dims <= 0:
+        raise ValueError("Volume and number of dimensions must be positive integers.")
+
+    # Step 1: Get the prime factors of vol
+    factors = prime_factors(vol)
+
+    # Step 2: Start with dimensions initialized to 1
+    shape = [1] * num_dims
+
+    # Step 3: Randomly assign factors to dimensions
+    for factor in factors:
+        random_dim = rng.randint(0, num_dims - 1)
+        shape[random_dim] *= factor
+
+    # Step 4: Ensure the final shape has the correct volume
+    if math.prod(shape) != vol:
+        raise ValueError("Generated shape does not match the desired volume.")
+
+    # If we get the same shape as source_shape, we try to add one dimension with size 1
+    if tuple(shape) == source_shape and len(shape) < max_dims:
```

Comment:
Can we leave output shape as it is if it turns out that it is the same to input shape? So we would have some small percentage of identity reshapes...

Reply:
@dgolubovicTT Thanks! Updated to keep the output shape if it matches the input.

Reply:
Identity shapes happened rarely, I thing only one of shapes that we tested generated same shape for output, but logging doesn't hurt.